### PR TITLE
refactor(android): optimize ImageView image download peformance

### PR DESCRIPTION
**Pull request by jquick-axway**
**JIRA:**
- https://jira.appcelerator.org/browse/TIMOB-28538
- https://jira.appcelerator.org/browse/TIMOB-18786

**Summary:**
- [TIMOB-28538](https://jira.appcelerator.org/browse/TIMOB-28538) Optimized `ImageView` image url download performance when:
  * HTTP failure response occurs.
  * When image is not cached. Happens when HTTP response contains `No-Cache`.
- [TIMOB-18786](https://jira.appcelerator.org/browse/TIMOB-18786) Fixed rare bug where wrong image is sometimes loaded due to hash code collision.
  * Now uses new `TiDrawableReference.Key` class which guarantees uniqueness.
  * No longer uses hash code as a key in a hash table.
- Removed `TiImageLruCache` and replaced with `TiImageCache`.
  * Old class kept hard references to bitmaps preventing large photos from being garbage collected.
  * New cache keeps soft references to all bitmaps allowing the Java garbage collector to better handle this when running low on heap memory.
  * New cache stores EXIF orientation. Prevents the need to read image binary more than once.
- Made singleton creation to be thread safe in several classes.
  * Fixed `TiDownloadManager`, `TiLoadImageManager`, `TiFileHelper`, and `TiBlobLruCache`.
  * Was wrongly creating multiple instances of `TiDownloadManager` on startup if there were several ImageViews on screen. Can cause performance initial performance issues since thread pool size would be ignored.
- Changed `ImageView` "decodeRetries" property default from `5` to `2`.

---
**ListView Complex Template Test:**
1. Build and run `ListViewTemplateComplexTest.js` attached to [TIMOB-28538](https://jira.appcelerator.org/browse/TIMOB-28538) on Android.
2. Scroll up and down in the ListView.
3. Verify it scrolls quickly. _(Might stutter initially but will be smooth from there.)_
4. Note that there will be 1 image missing every several rows and will log HTTP response errors. _(This is okay and verifies HTTP error responses are better handled now.)_

---
**ListView No-Cache Test:**
1. Build and run `ListViewImageRemoteNoCacheTest.js` attached to [TIMOB-28538](https://jira.appcelerator.org/browse/TIMOB-28538) on Android.
2. Scroll up and down in the ListView and verify it now scrolls quickly.
3. Verify all images are shown in portrait form. _(These images have various EXIF orientations.)_
4. Note that it might take a while to load the images since it involves HTTP redirects, but list should still scroll quickly.

---
**Image Hash Code Collision Test:**
1. Copy `FP.jpg` and `G1.jpg` attached to [TIMOB-18786](https://jira.appcelerator.org/browse/TIMOB-18786) to project's `./assets/images` folder.
2. Use `ImageHashTest.js` attached to [TIMOB-18786](https://jira.appcelerator.org/browse/TIMOB-18786) as your "app.js".
3. Build and run on Android.
4. Verify the 2 images displayed are different. _(Used to show the same image.)_

### CODE REVIEW

File: android/titanium/src/java/org/appcelerator/titanium/util/TiLoadImageManager.java [hunk 1]
```diff
@@ -44,106 +39,38 @@ public static TiLoadImageManager getInstance()
 		return _instance;
 	}
 
-	protected TiLoadImageManager()
+	private TiLoadImageManager()
 	{
-		handler = new Handler(this);
-		threadPool = Executors.newFixedThreadPool(THREAD_POOL_SIZE);
+		handler = new Handler(Looper.getMainLooper());
+		threadPool = Executors.newFixedThreadPool(4);
```
Comment (garymathews) [inline android/titanium/src/java/org/appcelerator/titanium/util/TiLoadImageManager.java:45]: We could either
- Define number of cores available on the device here
- Use [newCachedThreadPool()](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Executors.html#newCachedThreadPool) to dynamically scale on workload

**Review by garymathews — COMMENTED** [review]
Although unrelated to your changes, I'm seeing that _TiDrawableReference_ is retrying a bad image URL up to seven times. There's old code that does this due to a bug from Android 2.3 😱 

We should change `DEFAULT_DECODE_RETRIES` to `2`, amend [this](https://github.com/appcelerator/titanium_mobile/blob/2a878a29d83be85f92b060f8edc70148d152eb8b/android/titanium/src/java/org/appcelerator/titanium/view/TiDrawableReference.java#L412) `for` to:
```Java
for (int i = 1; i < decodeRetries; i++) {
```
and increase [this](https://github.com/appcelerator/titanium_mobile/blob/2a878a29d83be85f92b060f8edc70148d152eb8b/android/titanium/src/java/org/appcelerator/titanium/view/TiDrawableReference.java#L418) timeout to 1000ms or 3000ms

File: android/titanium/src/java/org/appcelerator/titanium/util/TiLoadImageManager.java (hunk 1 above)
Comment (jquick-axway) [reply android/titanium/src/java/org/appcelerator/titanium/util/TiLoadImageManager.java]: I changed it to use the number of CPU cores available. I did the same in `TiDownloadManager`.

**Review by garymathews — APPROVED** [review]
CR: **PASS**
FT: **PASS**

Here's a modified _ListViewImageRemoteNoCacheTest.js_ test so you can verify the images that are loaded are not the same.
```JS
const imageUrls = [
	'https://github.com/appcelerator/titanium_mobile/raw/master/tests/Resources/ExifFlipHorizontal.jpg',
	'https://github.com/appcelerator/titanium_mobile/raw/master/tests/Resources/ExifFlipVertical.jpg',
	'https://github.com/appcelerator/titanium_mobile/raw/master/tests/Resources/ExifRotate180.jpg',
	'https://github.com/appcelerator/titanium_mobile/raw/master/tests/Resources/ExifRotate270.jpg',
	'https://github.com/appcelerator/titanium_mobile/raw/master/tests/Resources/ExifRotate90.jpg',
	'https://github.com/appcelerator/titanium_mobile/raw/master/tests/Resources/ExifTranspose.jpg',
	'https://github.com/appcelerator/titanium_mobile/raw/master/tests/Resources/ExifTransverse.jpg',
];
function createListViewSectionItems() {
	const items = [];

	for (let index = 1; index <= 1000; index++) {
		const url = imageUrls[index % imageUrls.length];

		items.push({ properties: {
			title: `Row ${index}: ${url.split('/')[9]}`,
			image: url
		}});
	}
	return items;
}
const window = Ti.UI.createWindow();
window.add(
	Ti.UI.createListView({
		templates: {
			template: {
				properties: { layout: 'vertical' },
				childTemplates: [
					{
						type: 'Ti.UI.ImageView',
						bindId: 'image',
						properties: { width: 160 }
					},
					{
						type: 'Ti.UI.Label',
						bindId: 'title',
						properties: {
							color: 'black',
							font: { fontSize: '20dp', fontWeight:'bold' }
						}
					}
				]
			}
		},
		defaultItemTemplate: 'template',
		sections: [
			Ti.UI.createListSection({
				headerTitle: 'ListView',
				items: createListViewSectionItems(),
			})
		],
	})
);
window.open();
```

Also verified _error_ event still triggers, as this test was failing on Jenkins.
```JS
const win = Ti.UI.createWindow();
const img = Ti.UI.createImageView();

img.addEventListener('error', e => {
	alert(`Error loading image:\n\n${e.image}`);
});
img.image = 'https://invalid.host.com/image.jpg';

win.add(img);
win.open();
```

**Comment by lokeshchdhry** [comment]
FR Passed.

**Pull request merged**